### DISCUSSION TITLE
Is there a good git workflow to ease bringing in core changes?

### DISCUSSION THREAD

**Discussion by neekfenwick** [zencart/zencart#5934] [Q&A]
[This is probably a very old conversation, I wonder if anything has changed to make this easier]

Since `zencart` is developed using `git`, it seems there ought to be a fairly straightforward way to bring in latest merged changes from `zencart` `v158` branch (or whichever version your shop is based on) into one's own local repo [**edit** clarification: I mean a live shop's code, with significant long term changes on a feature branch that you always want to keep current against upstream... not just a clone of v158 for dev purposes].  However, the official docs say that to use Zen Cart we should download a tarball, but this would have no link back to the `zencart` git repo.  If you go the tarball route, you cannot use `git` tools to bring in changes from the source `git` repo.  I've generally found this makes my copy of zencart completely detached, because bringing in any changes is just too fiddly.

To try to avoid this, for example, to begin a new shop, one could clone `zencart`, checkout `v158` and create a new branch for your own shop:

```
$ git clone git@github.com:zencart/zencart.git myshop
$ cd myshop
$ git checkout -b myshop_master
```

One can then commit onto `myshop_master`, and to come up to speed with core development, we would:

```
$ git fetch
$ git rebase v158
```

This applies all v158 changes since we branched and puts our `myshop_master` branch on top, as if we had branched from the latest commit.

Do any of you guys use this workflow to keep up to date?  Is this doomed to failure down the road in some way I cannot see?

Does the alternative of using a tarball release, and cherry picking mainline `zencart` commits into it, work?  I would expect to run into problems where a `v158` commit relies on a previous commit, so if you don't cherry pick them in perfect and complete order then you run into horrible merge conflicts.  The point of wanting to work on a branch off `v158` is so that the `rebase` operation avoids any chance or forgetting to bring in commits from `v158`.

I'm experimenting today with trying to clone `zencart` fresh, add my dev repo as a remote, and cherry pick my changes into the fresh clone.  The immediate major hurdle is that we've renamed the `admin` folder to something else, so any changes happening to `admin` files in the `zencart` repo do not match filenames in my dev branch, e.g. `admin/index.php` and `ouradmin/index.php` are separate objects in `git`.  I can't see any way around this.

**Comment by lat9** [comment]
Simply add the following definition to your `admin/configure.php` (or `admin/local/configure.php`, if present):
```php
define('ADMIN_BLOCK_WARNING_OVERRIDE', '1');
```
That will enable you to continue using an admin-directory name of (er) `admin` for your testing purposes.

**Comment by neekfenwick** [comment]
> Simply add the following definition to your `admin/configure.php` (or `admin/local/configure.php`, if present):

Thanks @lat9 yes that's a great tip, I do this now and then, but the problem here is that the shop I'm deploying in the real world has a renamed `admin` folder.  So, a PR is merged changing `admin/index.php`, then `git pull` won't update my files because in my repo that file has been renamed to `myshop/index.php` (or whatever).

I could try renaming our admin folder back to `admin` in the local repo, and have our deploy step put `admin/*` into `remote:www_site/myshop` instead of `remote:www_site/admin`.  Then all my dev files are honestly `admin/*` and play nice with the origin repo.

**Reply by drbyte** [reply]
Ya, if your deploy script can do that, that's probably the simplest way.

Symlinking could be another option, but that defeats the (minor) benefit of obscuring the admin URL.

**Comment by neekfenwick** [comment]
After a week of leaving this question up I'm curious that no-one has responded either yes or no to a central question: do any of the people here use `git` to manage their real world shop code base, and if so, how do you manage bringing in core changes to that code base?

Has no-one tried it and has any good or bad stories to tell of how it works out?

I have a lot of respect for the work done over the years to enable modifications to Zen Cart to be handled outside of the core files, with the observer/notifier, template override and lang strings systems, but modifications can still be fairly messy and I'm hoping that `git`'s advanced merge algorithms can ease the pain of trying to keep a modified Zen Cart shop up to date.  I'm adding One Page Checkout to our store and every time I modify a core file with our customisations I wince :)

(I'm still open to trying it but is a slow burn project.. my current filesystem has zen cart as a physical copy of `v158` files from November in a subfolder of my `git` repo, with my commits touching both Zen Cart files and files outside of Zen (mostly a database changelog system using Liquibase), so bringing my commit history into a situation where the zencart repo is properly represented as a `git` repo is going to take some inventive submodule and `git patch` work - not how I would have structured things if I had planned to maintain the `zencart` upstream repo connection)

**Comment by drbyte** [comment]
I do use git for managing various client sites, but it's completely disconnected from the zencart/zencart repo. 
Whenever I do an upgrade I always do a 3-way compare between old-zc-vs-new-zc to identify changes and then copy those changes to the live site as the 3rd dimension of the merge (I use Araxis Merge for 3-way compares). I do that for core files, as well as all overridden files, and for any custom files that were cloned from something in core, to ensure anything that was updated in core gets applied anywhere else it needs to be ..... all done manually but with automated tools to reduce human error as much as possible. Same with any plugin files that might need adjusting for compatibility with the upgrade.
Then I commit the upgrade to git. And between upgrades of course I commit any customizations that are made, plugins installed/removed, etc.
So ... it's not really like what you initially asked about.

Now, would I "like" to be able to merge from the core branch on-demand? Sure there'd be times when that would be very convenient, like applying bugfixes-to-date, kinda constantly keeping the site as bleeding-edge (definitely would be helpful for battle-testing pre-release code). I do that by-hand presently, which in the end actually makes upgrades faster cuz I've been applying small patches all along the way. Of course, I don't recommend that for typical storeowners who aren't coders and comfortable studying potential side-effects of such an approach. But the geek in me doesn't want to wait for some updates, so I use them whenever I feel they'll be beneficial. (In truth, a lot of the code in Zen Cart has come from actual coding/features/fixes we've done for live sites.) But, I've digressed from your original topic again.

So ya, the pain-points you mentioned are real.
